working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+/// then any number of those registrations may trigger their associated
+/// callback. It will not necessarily be all or none of them.
+///
+/// Finalization callbacks will happen as *events*, not during execution of
+/// other code and not as a microtask, but as high-level events similar to
+/// timer events.
+abstract class FinalizationRegistry<F extends Object, FT> {
+  /// Creates a finalization registry with the given finalization callback.
+  external factory FinalizationRegistry(
+    void Function(FT finalizationToken) callback);
+
+  /// Registers [value] for a finalization callback.
+  ///
+  /// When [value] is no longer accessible to the program,
+  /// the registry *may* call its callback function with [finalizationToken]
+  /// as argument.
+  ///
+  /// The [value] and [unregisterToken] arguments do not count towards those
+  /// objects being accessible to the program. Both must be non-[num],
+  /// non-[String], non-[bool], non-[Pointer], non-[Struct] and
+  /// non-[Null] values.
+  ///
+  /// Multiple objects may be registered with the same finalization token,
+  /// and the same object may be registered multiple times with different,
+  /// or the same, finalization token.
+  ///
+  /// The callback may be called at most once per registration, and not
+  /// for registrations which have been unregistered since they were registered.
+  void register(F value, FT finalizationToken, {Object? unregisterToken});
+
+  /// Unregisters any finalization callbacks registered with [unregisterToken]
+  /// as unregister-token.
+  ///
+  /// After unregistering, those callbacks will not happen even if the
+  /// registered object becomes inaccessible.
+  void unregister(Object? unregisterToken);
+}
+
+/// Represents a weak reference to the given [target].
+///
+/// Weak references don't prevent garbage collection from reclaiming an object
+/// which is not reachable through any other (non-[WeakRef]) objects. If
+/// runtime system reclaims such _weakly reachable_ object then all [WeakRef]
+/// objects which reference it will be cleared to contain [null] as their
+/// target.
+abstract class WeakRef {
+  /// Create a [WeakRef] pointing to the given [target].
+  external factory WeakRef(Object? target);
+
+  /// Get the current object pointed by the [WeakRef].
+  external Object? get target;
+
+  /// Set the current object pointed by the [WeakRef].
+  external void set target (Object? newTarget);
+}
+```
+
+The following classes are added to `dart:ffi` library:
+
+```dart
+/// Any variable which has a static type that is a subtype of a [Finalizable]
+/// is guaranteed to be alive for the full duration of a scope in which it is
+/// defined.
+///
+/// In other words if an object is referenced by such a variable it is
+/// guaranteed to *not* be considered unreachable for the duration of the scope.
+abstract class Finalizable {
+  external factory Finalizable._();
+}
+
+typedef NativeFinalizer = Void Function(Pointer<Void>);
```

Comment:
This will not work on Windows. On Windows freeing is done with:

```dart
typedef WinHeapFreeNative = Int32 Function(
    Pointer heap, Uint32 flags, Pointer memory);
    
winHeapFree(processHeap, /*flags=*/ 0, pointer)
```

https://github.com/dart-lang/ffi/blob/db710989e4f9904cfc1c784f25e72030d7098151/lib/src/allocation.dart#L38-L39

Reply:
Yep, subsequent text highlights that.